docs/internal/generics.md

```diff
@@ -0,0 +1,114 @@
+# Generics
+
+Generics are a way to use type parameters to define or restrict other types.
+For example, a generic class can use its type parameters to define the type of fields, and a generic method can use its parameters to restrict what can be passed as an argument.
+
+Example:
+```ts
+class MyGenericType[MyType, SomeInterface]
+{
+  // MyType is the type of this field
+  var field : MyType;
+
+  // SomeInterface is the super type of anything that can be passed to this method
+  method(arg : SomeInterface)
+}
+
+// Some concrete types that implements Hashable
+class MyClass : Hashable { ... }
+class MyOtherClass : Hashable { ... }
+
+// This creates a concrete type
+let x : MyGenericType[U32, Hashable] & mut = MyGenericType;
+
+x.field = 42; // OK: The type of the field is compatible with integral literals
+
+let c = MyClass;
+x.method(c); // OK: MyClass implements Hashable
+let o = MyOtherClass;
+x.method(o); // OK: MyOtherClass also implements Hashable
+
+// This creates another concrete type, now with a more restrictive argument
+let y : MyGenericType[F64, MyClass] = MyGenericType;
+
+y.field = 42; // ERROR: No known conversion between integer literals and float types
+
+y.method(c); // OK: This is the expected type
+y.method(o); // ERROR: Even though MyOtherType implements the same interface, it's not he same type
+```
+
+In the example above, there is one generic class called `MyGenericClass` and two implementations (or specialisations) with types `[U32, Hashable]` and `[F64, MyClass]`.
+If all types are used correctly (ex. removing the ERROR lines above), the compiler will create two concrete classes from the one generic class: one for each specialisation.
+
+## Type Parameters
+
+Type parameters are the named types that can be used inside a generic declaration in place of actual types.
+In the example above, the generic class `MyGenericClass` has two type parameters: `MyType` and `Hashable`.
+Those names are used inside the class in place of existing (declared) types.
+In each specialisation, the type arguments (ex. `[U32, Hashable]`) replace the named arguments and the specialisation types are matched 1:1.
+
+Type parameters have the following properties:
+* A name: The name that will be used in the generic class as a substitute for the specialisation's passed declared name.
+* A restriction: What are the requirements (ex. interfaces, super-types, capabilities) of the type being passed in each specialisation.
+* A default type: In case the type is omitted from the specialisation declaration, a default type is assumed.
```

Comment:
```suggestion
* An optional default type: In case the type is omitted from the specialisation declaration, a default type is assumed.
  If no default is provided, a type must be provided at specialisation.
```